File: lib/token_lens/commands/render.rb
# frozen_string_literal: true

require "token_lens/parser"
require "token_lens/renderer/reshaper"
require "token_lens/renderer/annotator"
require "token_lens/renderer/layout"
require "token_lens/renderer/html"
require "token_lens/session"

module TokenLens
  module Commands
    class Render
      def initialize(output:, file_path: nil)
        @file_path = file_path
        @output = output
      end

      def run
        path = resolve_path
        warn "Rendering #{path}"
        tree = Parser.new(file_path: path).parse
        tree = Renderer::Reshaper.new.reshape(tree)
        tree.sort_by! { |n| n[:token].timestamp || "" }
        Renderer::Annotator.new.annotate(tree)
        canvas_width = Renderer::Layout.new.layout(tree)
        html = Renderer::Html.new(canvas_width: canvas_width).render(tree)
        File.write(@output, html)
        warn "Wrote #{@output}"
      end

      private

      def resolve_path
        return @file_path if @file_path
        sessions = Pathname.new(Dir.home).join(".token-lens", "sessions")
        saved = sessions.glob("*.json").max_by(&:mtime)
        return saved if saved
        warn "No saved captures found — reading active Claude Code session directly"
        Session.active_or_latest_jsonl
      end
    end
  end
end

File: src/commands/render.ts
import { writeFileSync } from "node:fs";
import { parse } from "../parser";
import { Html } from "../renderer/html";
import { annotate, layout } from "../renderer/layout";
import { reshape } from "../renderer/reshaper";
import { resolveSessionPath } from "./pick-session";

export interface RenderOptions {
  filePath?: string;
  output: string;
}

export async function render(opts: RenderOptions): Promise<void> {
  try {
    return await _render(opts);
  } catch (e) {
    process.stderr.write(`Error: ${e instanceof Error ? e.message : String(e)}\n`);
    process.exit(1);
  }
}

async function _render(opts: RenderOptions): Promise<void> {
  const filePath = await resolveSessionPath(opts.filePath);

  process.stderr.write(`Rendering ${filePath}\n`);
  const tree = parse(filePath);
  const reshapedTree = reshape(tree);
  if (reshapedTree.length === 0) {
    process.stderr.write(
      `Warning: no conversation turns found in ${filePath} — output will be empty.\n`,
    );
  }
  reshapedTree.sort((a, b) => (a.token.timestamp ?? "").localeCompare(b.token.timestamp ?? ""));
  annotate(reshapedTree);
  const canvasWidth = layout(reshapedTree);
  const html = new Html(canvasWidth).render(reshapedTree);
  writeFileSync(opts.output, html);
  console.log(`Wrote ${opts.output}`);
}
